src/devices/machine/spifi3.cpp

```diff
@@ -0,0 +1,1882 @@
+	const int cmd_entry = offset / 16;
+	uint8_t result = 0;
+	const int register_offset = offset % 16;
+	if (register_offset < 12)
+	{
+		result = spifi_reg.cmbuf[cmd_entry].cdb[register_offset];
+	}
+	else if (register_offset == 12)
+	{
+		result = spifi_reg.cmbuf[cmd_entry].quecode;
+	}
+	else if (register_offset == 13)
+	{
+		result = spifi_reg.cmbuf[cmd_entry].quetag;
+	}
+	else if (register_offset == 14)
+	{
+		result = spifi_reg.cmbuf[cmd_entry].idmsg;
+	}
+	else if (register_offset == 15)
+	{
+		result = spifi_reg.cmbuf[cmd_entry].status;
+	}
+
+	LOGMASKED(LOG_CMD, "cmd_buf_r(0x%x) -> 0x%x\n", offset, result);
+
+	return result;
+}
+
+void spifi3_device::cmd_buf_w(offs_t offset, uint8_t data)
+{
+	LOGMASKED(LOG_CMD, "cmd_buf_w(0x%x, 0x%x)\n", offset, data);
+
+	// 8 slots in the buffer, 16 bytes each
+	// so, divide the offset by 16 (truncated) to get the cmd entry
+	const int cmd_entry = offset / 16;
+	const int register_offset = offset % 16;
+	if (register_offset < 12)
+	{
+		spifi_reg.cmbuf[cmd_entry].cdb[register_offset] = data;
+	}
+	else if (register_offset == 12)
+	{
+		spifi_reg.cmbuf[cmd_entry].quecode = data;
+	}
+	else if (register_offset == 13)
+	{
+		spifi_reg.cmbuf[cmd_entry].quetag = data;
+	}
+	else if (register_offset == 14)
+	{
+		spifi_reg.cmbuf[cmd_entry].idmsg = data;
+	}
+	else if (register_offset == 15)
+	{
+		spifi_reg.cmbuf[cmd_entry].status = data;
+	}
+}
+
+TIMER_CALLBACK_MEMBER(spifi3_device::tick)
+{
+	step(true);
+}
+
+void spifi3_device::check_irq()
+{
+	// There are various ways interrupts can be triggered by the SPIFI - this method is a work in progress.
+	// TODO: ICOND, which doesn't seem to be needed much on the "happy path" (no errors)
+	const bool irqState = spifi_reg.intr & ~spifi_reg.imask;
+	if (irq != irqState)
+	{
+		LOGMASKED(LOG_INTERRUPT, "Setting IRQ line to %d\n", irqState);
+		irq = irqState;
+		m_irq_handler(irq);
+	}
+}
+
+void spifi3_device::check_drq()
+{
+	bool drq_state = drq;
+
+	switch (dma_dir)
+	{
+		case DMA_NONE:
+		{
+			drq_state = false;
+			break;
+		}
+
+		case DMA_IN: // device to memory
+		{
+			drq_state = !transfer_count_zero() && !m_even_fifo.empty();
+			break;
+		}
+
+		case DMA_OUT: // memory to device
+		{
+			drq_state = !transfer_count_zero() && !m_even_fifo.full();
+			break;
+		}
+	}
+
+	if (drq_state != drq)
+	{
+		LOGMASKED(LOG_DATA, "DRQ changed to %d!\n", drq_state);
+		drq = drq_state;
+		m_drq_handler(drq);
+	}
+}
+
+bool spifi3_device::transfer_count_zero()
+{
+	return spifi_reg.icond & ICOND_CNTZERO;
+}
+
+void spifi3_device::reset_disconnect()
+{
+	scsi_bus->ctrl_w(scsi_refid, 0, ~S_RST);
+
+	command_pos = 0;
+	mode = MODE_D;
+}
+
+void spifi3_device::send_byte(scsi_data_target data_source)
+{
+	state = (state & STATE_MASK) | (SEND_WAIT_SETTLE << SUB_SHIFT);
+
+	if (data_source == COMMAND_BUFFER)
+	{
+		// Send next data from cmbuf
+		if (command_pos > 11)
+		{
+			fatalerror("Tried to send command past the end of cdb! Command_pos: %d", command_pos);
```

Comment:
You should include the device tag in the message for fatal errors to give context (unlike `logerror`, the tag isn’t automatically included).